File: Sources/NotchleMac/UI/NotchViewModel.swift
import Foundation
import Observation
import NotchleCore

/// The one object the notch UI binds to. Owned by the integrator (frozen contract):
/// views read `state`/`settings` and call `send`/`updateSettings`; the app coordinator
/// fills in the closures and pushes new state after every engine step.
@MainActor
@Observable
public final class NotchViewModel {
    public var state: GameState
    public var settings: AppSettings
    /// Name and capability of the active player, for the settings sheet and the
    /// "preview ends after 30s" hint.
    public var playerName: String = ""
    public var playerPlaysFullTrack: Bool = true
    /// Spotify Connect sign-in state (additive; nil in demos and tests that don't need it).
    public var spotifyConnect: SpotifyConnectModel?
    /// Every recorded track, in the order played (agreed addition, 2026-09-24). Views never
    /// show this raw: they go through `HistorySpoilerFilter.visible(_:in:)`.
    public var history: [HistoryEntry] = []
    /// Album cover of the current track, set only once its answer is on screen (correct or
    /// revealed) and cleared when the phase moves on. Views still gate it through
    /// `NotchUIRules.revealedArtworkURL(_:_:)`.
    public var currentArtworkURL: URL?

    @ObservationIgnored public var send: (GameAction) -> Void = { _ in }
    @ObservationIgnored public var updateSettings: (AppSettings) -> Void = { _ in }
    @ObservationIgnored public var clearHistory: () -> Void = {}

    public init(state: GameState = GameState(), settings: AppSettings = AppSettings()) {
        self.state = state
        self.settings = settings
    }
}

File: windows/src/Notchle.Windows/NotchViewModel.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;
using Notchle.Core;

namespace Notchle.Windows;

/// The one object the island UI binds to (frozen contract, owned by the integrator).
/// Views read State/Settings and call Send/UpdateSettings; the coordinator sets the delegates
/// and assigns new State on the UI thread after every engine step.
public sealed class NotchViewModel : INotifyPropertyChanged
{
    private GameState _state = new() { Config = GameConfig.Default };
    private AppSettings _settings = new();
    private string _playerName = "";
    private bool _playerPlaysFullTrack;

    public GameState State { get => _state; set => Set(ref _state, value); }
    public AppSettings Settings { get => _settings; set => Set(ref _settings, value); }
    public string PlayerName { get => _playerName; set => Set(ref _playerName, value); }
    public bool PlayerPlaysFullTrack { get => _playerPlaysFullTrack; set => Set(ref _playerPlaysFullTrack, value); }

    public Action<GameAction> Send { get; set; } = _ => { };
    public Action<AppSettings> UpdateSettings { get; set; } = _ => { };

    public event PropertyChangedEventHandler? PropertyChanged;

    private void Set<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value)) return;
        field = value;
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
    }
}
